security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,188 @@
+	caSecret *v1.Secret) {
+	if caSecret == nil {
+		rootCertRotatorLog.Errorf("root cert secret %s is nil, skip cert rotation job",
+			CASecret)
+		return
+	}
+	// Check root certificate expiration time in CA secret
+	waitTime, err := rotator.config.certInspector.GetWaitTime(caSecret.Data[caCertID], time.Now(), time.Duration(0))
+	if err == nil && waitTime > 0 {
+		rootCertRotatorLog.Info("Root cert is not about to expire, skipping root cert rotation.")
+		caCertInMem, _, _, _ := rotator.ca.GetCAKeyCertBundle().GetAllPem()
+		// If CA certificate is different from the CA certificate in local key
+		// cert bundle, it implies that other Citadels have updated istio-ca-secret.
+		// Reload root certificate into key cert bundle.
+		if !bytes.Equal(caCertInMem, caSecret.Data[caCertID]) {
+			rootCertRotatorLog.Warn("CA cert in KeyCertBundle does not match CA cert in " +
+				"istio-ca-secret. Start to reload root cert into KeyCertBundle")
+			if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(caSecret.Data[caCertID],
+				caSecret.Data[caPrivateKeyID], nil, caSecret.Data[caCertID]); err != nil {
+				rootCertRotatorLog.Errorf("failed to reload root cert into KeyCertBundle (%v)", err)
+			}
+			rootCertRotatorLog.Info("Successfully reloaded root cert into KeyCertBundle.")
+		}
+		return
+	}
+
+	rootCertRotatorLog.Infof("Refresh root certificate, root cert is about to expire: %s", err.Error())
+	options := util.CertOptions{
+		TTL:           rotator.config.caCertTTL,
+		SignerPrivPem: caSecret.Data[caPrivateKeyID],
+		Org:           rotator.config.org,
+		IsCA:          true,
+		IsSelfSigned:  true,
+		RSAKeySize:    caKeySize,
+		IsDualUse:     rotator.config.dualUse,
+	}
+	pemCert, pemKey, ckErr := util.GenRootCertFromExistingKey(options)
+	if ckErr != nil {
+		rootCertRotatorLog.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+		return
+	}
+
+	caSecret.Data[caCertID] = pemCert
+	caSecret.Data[caPrivateKeyID] = pemKey
+	if err = rotator.caSecretController.UpdateCASecretWithRetry(caSecret, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write secret to CA secret (error: %s). "+
+			"Abort new root certificate.", err.Error())
+		return
+	}
+	rootCertRotatorLog.Infof("A new self-generated root certificate is written into secret: %v", string(pemCert))
+
+	if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(pemCert, pemKey, nil, pemCert); err != nil {
+		rootCertRotatorLog.Errorf("failed to create CA KeyCertBundle (%v)", err)
+		return
+	}
+
+	certEncoded := base64.StdEncoding.EncodeToString(rotator.ca.GetCAKeyCertBundle().GetRootCertPem())
+	if err = rotator.configMapController.InsertCATLSRootCertWithRetry(
+		certEncoded, rotator.config.retryInterval, 30*time.Second); err != nil {
+		rootCertRotatorLog.Errorf("Failed to write self-signed Citadel's root cert "+
+			"to configmap (%s). Node agents will not be able to connect.",
+			err.Error())
+	}
```

Comment:
Copy this code to line#45 to periodically sync the root cert in ConfigMap with the root cert in secret.

Reply:
Done. Thanks.